teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+- [Alternatives](#alternatives)
+- [Upgrade &amp; Migration Strategy](#upgrade--migration-strategy)
+- [Implementation Pull request(s)](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes expanding support for arrays by adding:
+* Support for indexing into arrays in variable replacement
+  (in addition to [replacing entire lists](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#substituting-array-parameters))
+* Support for array [results](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results) (for both Pipelines and Tasks)
+
+## Motivation
+
+* See [use cases](#use-cases) - use cases for looping support in pipelines often need an array input;
+  being able to produce an array from a Task to feed into a loop would enable some interesting CD use cases
+* TEP-0075 proposing adding dictionary support in addition to the existing support strings
+  and arrays as params, and results, but today we only support arrays as params and not as
+  results. The proposal to add dictionary is mostly motivated by use cases for dictionary
+  results (vs params) - but today we only allow string results. It seems confusingly
+  inconsistent to imagine adding dictionary results but not array results (even though
+  we support array params)
+* Without indexing, arrays will only work with arrays. This means you cannot combine arrays with Tasks that
+  only take string parameters and this limits the interoperabiltiy between Tasks.
+
+### Goals
+
+* Add two missing pieces of array support
+  * As results
+  * Indexing into arrays
+* Take a step in the direction of allowing Tasks to have even more control
+  over their params and results ([see pipelines#1393](https://github.com/tektoncd/pipeline/issues/1393))
+* Be consistent in our treatment of arrays in Pipelines and Tasks and other types, e.g.
+  dictionaries (TEP-0075)
+
+### Non-Goals
+
+* Adding complete [JSONPath](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html)
+  syntax support
+* Adding support for nesting, e.g. array params where the items in the array are themselves arrays
+  (no reason we can't add that later but trying to keep it simple for now)
+
+### Use Cases
+
+1. Looping (i.e. providing values for an interface such as the task-loop custom task
+   [iterateParam](https://github.com/tektoncd/experimental/tree/main/task-loops#specifying-the-iteration-parameter)
+   (see the workaround added in [experimental#713](https://github.com/tektoncd/experimental/pull/713) which
+   splits a string). For example, in tektoncd/plumbing we have a directory of
+   [images](https://github.com/tektoncd/plumbing/tree/main/tekton/images). Each directory contains a Dockerfile -
+   in a Pipeline that builds all images in this repo, we would want to: 1) determine which Dockerfiles need to be built
+   and 2) for each of those Dockerfiles, build the image. Looping support would allow us to use a Task which builds
```

Comment:
I would extend this by saying that we also need to be capture results from a loop.
It we build an array of container images, we need to be able to get back an arrays of results.
In general, it is not possible to use a task with results in a loop because of this today.

Reply:
Ahh interesting, this makes me think of 2 more use cases we could add (let me know if this is what you're thinking too or if I'm missing the point):

1. If the "loop" is inside of the Task, e.g. if a task knew how to take an array of dockerfile locations and build each of them, then I think I agree that you'd need to be able to produce an array of results in order to contain the results of each iteration (maybe i can add this as a use case and use our [release publish task](https://github.com/tektoncd/pipeline/blob/main/tekton/publish.yaml) as an example - it builds mulitple images AND it loops over them and publishes them to multiple locations resulting in multiple URLs)
2. It seems like it's important to let custom tasks (via Runs) emit array results because you don't know what kind of looping or other logic they might contain - e.g. [the existing looping custom task mentions the fact that it doesnt create results as a limitation](https://github.com/tektoncd/experimental/tree/main/task-loops#limitations) - and with the current interface, it would have to combine the results of each task into a string

(the last question would be what does the results interface look like if we support looping directly in the Tekton API but we can leave that for a looping focused TEP :D)

Reply:
or maybe a better way to state it, and maybe this is more what you were getting at @afrittoli, is supporting looping where you both need to **consume** array results (to loop over) and **produce** array results (from those loops)